lib/OpenQA/WebAPI/Plugin/AMQP.pm

```diff
@@ -69,6 +70,178 @@ sub log_event {
     # seperate function for tests
     $self->publish_amqp($topic, $event_data);
 }
+sub log_event_fedora_ci_messages {
+    # this is for emitting messages in the "CI Messages" format:
+    # https://pagure.io/fedora-ci/messages
+    # This is a Fedora/Red Hat-ish thing in a way, but in theory
+    # anyone could adopt it
+    my ($self, $event, $job, $baseurl) = @_;
+    my $stdevent;
+    my $clone_of;
+    my $job_id;
+    # first, get the standard 'state' (from 'queued', 'running',
+    # 'complete', 'error'; we cannot do 'running' at present
+    if ($event eq 'openqa_job_create') {
+        $stdevent = 'queued';
+        $job_id   = $job->id;
+    }
+    elsif ($event eq 'openqa_job_restart' || $event eq 'openqa_job_duplicate') {
+        $stdevent = 'queued';
+        $clone_of = $job->id;
+        $job_id   = $job->clone_id;
+    }
+    elsif ($event eq 'openqa_job_cancel') {
+        $stdevent = 'error';
+        $job_id   = $job->id;
+    }
+    elsif ($event eq 'openqa_job_done') {
+        $job_id = $job->id;
+        # lifecycle note: any job cancelled directly via the web API will
+        # see both job_cancel and job_done with result USER_CANCELLED, so
+        # we emit duplicate standardized fedmsgs in this case. This is
+        # kinda unavoidable, though, as it's possible for a job to wind up
+        # USER_CANCELLED *without* an openqa_job_cancel event happening,
+        # so we can't just throw away all openqa_job_done USER_CANCELLED
+        # events...
+        if (grep { $job->result eq $_ } OpenQA::Schema::Result::Jobs::INCOMPLETE_RESULTS) {
+            $stdevent = 'error';
+        }
+        else {
+            $stdevent = 'complete';
+        }
```

Comment:
I mean, sure, this is copied straight out of the fedmsg plugin, though. I just copied the whole thing and tweaked it a bit (for the updated spec).

Reply:
Using the constant like that is wrong though. It's coincidence that `OpenQA::Schema::Result::Jobs` has it imported. The real source is `OpenQA::Jobs::Constants`.

Reply:
well, I think it's rather that this was *changed* - this is copied from the fedmsg plugin as I said, and I think when I wrote that, `OpenQA::Jobs::Constants` didn't exist, it was invented later. since this is 'new' code I'll change it...

Reply:
for the record, this turned out to need brackets around the grep expression :)